### PR TITLE
remove deepcopying

### PR DESCRIPTION
**Pull request by josh146**

### CODE REVIEW

File: pennylane/tape/tapes/jacobian_tape.py [hunk 1]
```diff
@@ -533,6 +527,12 @@ def jacobian(self, device, params=None, **options):
                 # object arrays cannot be flattened; must hstack them
                 g = np.hstack(g)
 
-            jac[:, i] = g.flatten()
+            g = g.flatten()
+
+            if jac is None:
+                jac = np.zeros((len(g), len(params)), dtype=float)
+                self._output_dim = len(g)
```
Comment (mariaschuld) [inline pennylane/tape/tapes/jacobian_tape.py:534]: Clever!

File: pennylane/operation.py [hunk 2]
```diff
@@ -241,6 +241,18 @@ class Operator(abc.ABC):
     """
     do_check_domain = True  #: bool: flag: should we perform a domain check for the parameters?
 
+    def __copy__(self):
+        cls = self.__class__
+        copied_op = cls.__new__(cls)
+        copied_op.data = self.data.copy()
```
Comment (josh146) [inline pennylane/operation.py:247]: The key here is that the parameters are also shallow copied

File: pennylane/operation.py [hunk 3]
```diff
@@ -1169,6 +1181,13 @@ def __init__(self, *args):  # pylint: disable=super-init-not-called
             else:
                 raise ValueError("Can only perform tensor products between observables.")
 
+    def __copy__(self):
+        cls = self.__class__
+        copied_op = cls.__new__(cls)
+        copied_op.obs = self.obs.copy()
```
Comment (josh146) [inline pennylane/operation.py:1187]: Same here - we shallow copy all observables.

**Review by mariaschuld — APPROVED** [review]
Great, so you basically now have deepcopy copy the operations manually, which in turn just updates the correct elements of the operations manually - this stops the huge recursion, right?

As far as I can see this is perfect, and if there happens to be information that we forgot in the copy we will just have to find out over time.

**Comment by josh146** [comment]
> Great, so you basically now have deepcopy copy the operations manually, which in turn just updates the correct elements of the operations manually - this stops the huge recursion, right?

Yep! It's essentially a shallow copy of everything + also a shallow copy of all operations and observables.

File: pennylane/tape/measure.py [hunk 4]
```diff
@@ -76,6 +78,20 @@ def __init__(self, return_type, obs=None, wires=None, eigvals=None):
         # Queue the measurement process
         self.queue()
 
+    def __copy__(self):
+        cls = self.__class__
+        copied_op = cls.__new__(cls)
+
+        copied_op.return_type = self.return_type
```
Comment (co9olguy) [inline pennylane/tape/measure.py:85]: Why is `return_type` still hanging around for the constituent ops? (shouldn't all this logic have migrated wholy to the `MeasurementProcess`?

File: pennylane/tape/tapes/tape.py [hunk 5]
```diff
@@ -868,28 +868,21 @@ def copy(self, deep=False, tape_cls=None):
...
+        else:
+            tape._prep = self._prep.copy()
+            tape._ops = self._ops.copy()
+            tape._measurements = self._measurements.copy()
 
         tape._update()
-        tape._par_info = self._par_info.copy()
         tape.trainable_params = self.trainable_params.copy()
 
         tape._caching = self._caching
```
Comment (co9olguy) [inline pennylane/tape/tapes/tape.py:888]: No new tests??

File: pennylane/tape/measure.py (hunk 4 above)
Comment (josh146) [reply pennylane/tape/measure.py:85]: This is the `MeasurementProcess` :) I'll change the variable name

**Pull request merged**